### DISCUSSION TITLE
Verification, Password Reset and Email Change with OTP

### DISCUSSION THREAD

**Discussion by chnfyi** [pocketbase/pocketbase#5714] [Q&A] [answered]
Hi ! With v0.23.0.rc-7, i would like to use the new OTP feature to manage user email verification on signup without redirection - unlike email token from `requestVerification()` as path segment or query parameter.

```
// pb_hooks/main.pb.js

onMailerRecordVerificationSend((e) => {
   // Disable verification email
   return false
}, "users")
```

```
// Form first step

await pb.collection("users").create({
   email: "test@example.com",
   password: "123456",
   passwordConfirm: "123456"
})

// Form second step

const result = await pb.collection("users").requestOTP("test@example.com")
await pb.collection("users").authWithOTP(result.otpId, "010203") // Then user should be verified
```

But I have not found some workaround to use one-time code instead of email token for reset password and email change flows.

Is there a way to use OTP interface in a custom route to achieve such custom flows ?

Thanks.

**Comment by ganigeorgiev** [comment, marked as answer]
Based on my understanding of the title (_the description doesn't really make sense to me_) you can't use the builtin OTP for this, at least not out of the box. At the minimum you'll need also somewhere to store the type of what the OTP is because in your case it could be for more than just auth.

Without details about your actual use case and why what you are trying to do is even needed in the first place, it remains out of the scope of the project.

If your app requires a more custom auth flows then what we already have, I'd recommend to use something else as most likely you'll mess up the security if you try to force PocketBase into your use case.

**Reply by ganigeorgiev** [reply]
If you just want to mark the user as "verified" after OTP auth, then you can use something like:

```js
onRecordAuthWithOTPRequest((e) => {
   e.next()
   
   // mark the user as verified (if not already)
   // (optionally you can move it before the e.next() but note that OTP can fail if you have MFA enabled and this is the first auth)
   if (!e.record.get("verified")) {
      e.record.set("verified", true)
      e.app.save(e.record)
   }
}, "users")
```

**Reply by chnfyi** [reply]
My idea was to avoid links in verification, password reset and email change emails.

I wanted to use a code sent by email instead of a token when a user confirm password reset or confirm email change.

Like `confirmPasswordReset('EMAIL_CODE', '1234567890', '1234567890')` instead of `confirmPasswordReset('TOKEN', '1234567890', '1234567890')`

I know that it is not a PocketBase feature and i understand that, with a custom endpoint, i would need more than the built-in OTP and i would need to store which type of request (verification, password reset or email change) generated the OTP to make it work properly.

I will stay with the normal use case.

Thanks for your work and your reply !
